docs/replica-sets.md

```diff
@@ -0,0 +1,117 @@
+# VirtualMachineReplicaSet
+
+## Overview
+
+In order to allow scaling up or down similar VMs, add a way to specify
+`VirtualMachine` templates and the amount of replicas required, to let the
+runtime create these `VirtualMachine`s.
+
+## Use-cases
+
+### Ephemeral VMs
+
+Scaling ephemeral VMs which only have read-only mounts, or work with a backing
+store, to keep temporary data, which can be deleted after the VM gets
+undefined.
+
+## Implementation
+
+A new object `VirtualMachineReplicaSet`, backed with a controller will be
+implemented:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachineReplicaSet
+metadata:
+  name: myreplicaset
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      mylabel: mylabel
+  template:
+    metadata:
+      name: test
+      labels:
+        mylabel: mylabel
+    spec:
+      domain:
+        devices:
+      [...]  
+```
+
+`spec.template` is equal to a `VirtualMachineSpec`. `spec.replicas` specifies
+how many instances should be created out of `spec.temlate`. `spec.selector`
+contains selectors, which need to match `spec.template.metadata.labels`.
+
+The status looks like this:
+
+```yaml
+status:
+  conditions: null
+  replicas: 3
+```
+
+It shows the number of `VirtualMachine`s which are in a non-final state nad
+which match `spec.selector`. In case of a scaling error, a `ReplicaFailure`
+condition is added to the status.
+
+In case of a delete failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureDelete"
+    message: "no permission to delete VMs"
+    lastTransmissionTime: "..."
+  replicas: 4
+```
+
+In case of a create failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureCreate"
+    message: "no permission to create VMs"
+    lastTransmissionTime: "..."
+  replicas: 2
+```
+
+### Guarantees
+
+Once Graceful Deletes are implemented, the VirtualMachineReplicaSet guarantees
+that it will never create more than the requested numbers of VMs in a non-final
+state.
+
+As a consequence, in case of node failures, or in case of connection loss to
+virt-handlers which run VMs which are part of that set, no new VMs will be
+spawned until the VMs in unknown state are explicitly removed, or the node
+reconnects. This behaviour allows plugging in fencing controllers in a well
+defined way.
+
+It might make sense to weaken these guarantees in the future, and instead add a
+`StatefulSet` equivalent to KubeVirt in the future.
```

Comment:
From an implementation perspective i was going to recommend that you *not* recreate the entire controller flow yourself, and instead simply spawn a new RS or SS with your desired state via the controller.  If you do that, you'll be limited to only the guarantees those controllers provide, but you won't have to reimplement the entire stack.

If you are planning on replicating the code stack, add some justifications here as to why.

Reply:
> From an implementation perspective i was going to recommend that you not recreate the entire controller flow yourself, and instead simply spawn a new RS or SS with your desired state via the controller.

@smarterclayton You might be right. This is something we've been discussing quite a bit this week.

Once we begin looking at what is involved with recreating the SS controller, I think there's a good chance we'll want to leverage the existing k8s controller over creating the controller flow ourselves. We don't have a clear understanding of the complexity involved or if our use case diverges from the k8s SS/RS controllers yet.